packages/common/core-interfaces/src/exposedUtilityTypes.ts

```diff
@@ -0,0 +1,565 @@
+								T[K],
+								{ Replaced: Options["Replaced"] | T }
+							>;
+						}
+					: /* not array => error */ SerializationErrorPerNonPublicProperties
+				: /* no hidden properties => apply filtering => */ JsonSerializableFilter<
+						T,
+						Options["Replaced"]
+					>;
+
+	/**
+	 * Core implementation of {@link JsonSerializable}.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export type JsonSerializableFilter<T, TReplaced> = /* test for 'any' */ boolean extends (
+		T extends never
+			? true
+			: false
+	)
+		? /* 'any' => */ JsonTypeWith<TReplaced>
+		: /* test for 'unknown' */ unknown extends T
+			? /* 'unknown' => */ JsonTypeWith<TReplaced>
+			: /* test for JSON Encodable primitive types or given alternate */ T extends
+						| null
+						| boolean
+						| number
+						| string
+						| TReplaced
+				? /* primitive types => */ T
+				: // eslint-disable-next-line @typescript-eslint/ban-types
+					/* test for not a function */ Extract<T, Function> extends never
+					? /* not a function => test for object */ T extends object
+						? /* object => test for array */ T extends readonly (infer _)[]
+							? /* array => */ {
+									/* array items may not not allow undefined */
+									/* use homomorphic mapped type to preserve tuple type */
+									[K in keyof T]: JsonForSerializableArrayItem<
+										T[K],
+										TReplaced,
+										JsonSerializableFilter<T[K], TReplaced | T>
+									>;
+								}
+							: /* not an array => test for exactly `object` */ IsExactlyObject<T> extends true
+								? /* `object` => */ NonNullJsonObjectWith<TReplaced>
+								: /* test for enum like types */ IsEnumLike<T> extends true
+									? /* enum or similar simple type (return as-is) => */ T
+									: /* property bag => */ FlattenIntersection<
+											{
+												/* required properties are recursed and may not have undefined values. */
+												[K in keyof T as RequiredNonSymbolKeysOf<
+													T,
+													K
+												>]-?: undefined extends T[K]
+													? {
+															["error required property may not allow undefined value"]: never;
+														}
+													: JsonSerializableFilter<T[K], TReplaced | T>;
+											} & {
+												/* optional properties are recursed and allowed to preserve undefined value type. */
+												[K in keyof T as OptionalNonSymbolKeysOf<
+													T,
+													K
+												>]?: JsonSerializableFilter<T[K], TReplaced | T | undefined>;
+											} & {
+												/* symbol properties are rejected */
+												[K in keyof T & symbol]: never;
+											}
+										>
+						: /* not an object => */ never
+					: /* function => */ never;
+
+	/**
+	 * Sentinel type for use when marking points of recursion (in a recursive type).
+	 * Type is expected to be unique, though no lengths is taken to ensure that.
+	 *
+	 * @beta
+	 * @system
+	 */
+	export interface RecursionMarker {
+		"recursion here": "recursion here";
+	}
```

Comment:
It's possible for this to collide with regular data. Picking something faster to compare and collision resistant seems like it would be a good idea. unique symbols seem ideal for that.

Reply:
I was able to switch to symbols.